### PR TITLE
Add possibility to query for job results

### PR DESCRIPTION
**Pull request by sschober**
The `startCommand()` methods return opaque job ids. These commits introduce a new client methods `queryJobResult()`, which enable querying for the results of those jobs.

The `Connection` class was extended by one method, enabling the creation of `GET` requests.

### CODE REVIEW

File: src/main/java/com/suse/saltstack/netapi/client/SaltStackClient.java [hunk 1]
```diff
@@ -290,6 +290,37 @@ public Job call() throws SaltStackException {
...
+     * @return Map key: minion id, value: command result from that minion
+     * @throws SaltStackException 
+     */
+    public Map<String,Object> queryJobResult( final String job) throws SaltStackException{
+        Result<List<Map<String,Object>>> result = connectionFactory
+                .create("/jobs/"+job, JsonParser.RETVALS, config)
+                .getResult();
+
+        // A list with one element is returned, we take the first
+        return result.getResult().get(0);
```
Comment (isbm) [inline src/main/java/com/suse/saltstack/netapi/client/SaltStackClient.java]: What will happen if result gets empty? Maybe:
```java
    ...
    return (!result.getResult().isEmpty() ? result.getResult().get(0) : null)
```

...or new empty List (depends). But certainly I would be worried about get something like "index out of range" error.

File: src/main/java/com/suse/saltstack/netapi/client/impl/HttpClientConnection.java [hunk 2]
```diff
@@ -108,5 +143,5 @@ public T getResult(String data) throws SaltStackException {
         } catch (IOException e) {
             throw new SaltStackException(e);
         }
-    }
+	}
```
Comment (isbm) [inline src/main/java/com/suse/saltstack/netapi/client/impl/HttpClientConnection.java]: The formatting become a little bit original here.

File: src/main/java/com/suse/saltstack/netapi/client/impl/JDKConnection.java [hunk 3]
```diff
@@ -45,6 +45,14 @@ public T getResult(String data) throws SaltStackException {
     }
 
     /**
+     * {@inheritDoc}
+     */
+	@Override
+	public T getResult() throws SaltStackException {
+		return request("GET", null);
```
Comment (isbm) [inline src/main/java/com/suse/saltstack/netapi/client/impl/JDKConnection.java]: What do you use for formatting? It does not look like 4 spaces to me.

File: src/test/java/com/suse/saltstack/netapi/client/SaltStackClientTest.java [hunk 4]
```diff
@@ -196,4 +196,21 @@ public void testRunResultAsync() throws SaltStackException {
...
         assertEquals(retvals.get("minion-1"), true);
     }
+
+    @Test
+    public void testQueryJobResult() throws SaltStackException {
+        stubFor(get(urlEqualTo("/jobs/some-job-id"))
+                .willReturn(aResponse()
+                    .withStatus(HttpURLConnection.HTTP_OK)
+                    .withHeader("Content-Type", "application/json")
+                    .withBody(JSON_RUN_RESPONSE)));
```
Comment (isbm) [inline src/test/java/com/suse/saltstack/netapi/client/SaltStackClientTest.java]: This also could be formatted in one line under the `.willReturn...` method.

File: src/test/java/com/suse/saltstack/netapi/client/SaltStackClientTest.java [hunk 5]
```diff
@@ -196,4 +196,21 @@ public void testRunResultAsync() throws SaltStackException {
...
+
+        Map<String, Object> retvals = client.queryJobResult("some-job-id");
+
+        assertNotNull(retvals);
+        assertTrue(retvals.containsKey("minion-1"));
+        assertEquals(retvals.get("minion-1"), true);
+    }
+
+
```
Comment (isbm) [inline src/test/java/com/suse/saltstack/netapi/client/SaltStackClientTest.java]: Is there a reason for these two empty lines?

File: src/main/java/com/suse/saltstack/netapi/client/impl/HttpClientConnection.java [hunk 6]
```diff
@@ -55,6 +64,25 @@ public HttpClientConnection(String endpointIn, JsonParser<T> parserIn,
...
+	
+    /**
+     * Perform HTTP request and parse the result into a given result type.
+     *
+     * @param method the HTTP method to use
+     * @return object of type given by resultType
+     * @throws SaltStackException in case of a problem
+     */
+	private T request(String data) throws SaltStackException{
```
Comment (isbm) [inline src/main/java/com/suse/saltstack/netapi/client/impl/HttpClientConnection.java]: It has to be properly formatted too.

File: src/main/java/com/suse/saltstack/netapi/client/Connection.java [hunk 7]
```diff
@@ -8,7 +8,16 @@
...
+	
+	/**
+	 * Send a GET request and parse the result into object of given {@link Type}.
+	 * 
+	 * @return object of type given by resultType
+	 * @throws SaltStackException
+	 */
+	T getResult() throws SaltStackException;
+    
+	/**
```
Comment (isbm) [inline src/main/java/com/suse/saltstack/netapi/client/Connection.java]: Please fix formatting here.

File: src/main/java/com/suse/saltstack/netapi/client/impl/HttpClientConnection.java [hunk 8]
```diff
@@ -77,25 +105,32 @@ public T getResult(String data) throws SaltStackException {
...
-            // POST data
+            HttpUriRequest req = null;
             if (data != null) {
+                // POST data
+                HttpPost httpPost = new HttpPost(uri);
                 httpPost.addHeader("Content-Type", "application/json");
                 httpPost.setEntity(new StringEntity(data));
+                req = httpPost;
+            }
+            else {
```
Comment (isbm) [inline src/main/java/com/suse/saltstack/netapi/client/impl/HttpClientConnection.java]: Please always:
```java
} else {
```

**Comment by renner** [comment]
Hello and thanks for contributing! It is very much appreciated that you are proposing a patch to actually extend the functionality of the library. It would be very nice though if you could fix the formatting of your patch according to our project guidelines as well as regarding the inline comments. The most important formatting rules can be found in the [README] (https://github.com/SUSE/saltstack-netapi-client-java/blob/master/README.md), let me know if you have any questions.

Please note further that we agreed on a [convention about naming the library methods] (https://github.com/SUSE/saltstack-netapi-client-java/pull/15#issuecomment-74485966). While the currently proposed method name `queryJobResult()` is quite OK, I would propose to change it to `getJobResult()`.

File: src/main/java/com/suse/saltstack/netapi/client/SaltStackClient.java (hunk 1 above)
Comment (sschober) [inline src/main/java/com/suse/saltstack/netapi/client/SaltStackClient.java]: You are right, but I think the same reasoning should apply at the end of the `run()` and `startCommand()` methods. Is there a project policy about returning `null` vs empty list?

Comment (isbm) [inline src/main/java/com/suse/saltstack/netapi/client/SaltStackClient.java]: Well, it sounds like something you could propose and contribute this direction! :)

**Comment by moio** [comment]
Hey @sschober, thanks for this PR!

We like the content and could find no obvious problem with it, so we will gladly merge your contribution. 

There are still some smaller style issues here and there though, so I will add inline comments - in general my recommendation (even in other OSS projects) is to copy whatever style you see in surrounding code.

Thanks in advance!

File: src/main/java/com/suse/saltstack/netapi/client/SaltStackClient.java [hunk 9]
```diff
@@ -290,6 +290,37 @@ public Job call() throws SaltStackException {
     }
 
     /**
+     * Query for result of supplied job.
+     * 
+     * GET /job/<job-id>
+     * 
+     * @param job {@link Job} object representing scheduled job
+     * @return Map key: minion id, value: command result from that minion
+     * @throws SaltStackException
```
Comment (moio) [inline src/main/java/com/suse/saltstack/netapi/client/SaltStackClient.java]: Nitpick: add descriptive comments for Exceptions.

File: src/main/java/com/suse/saltstack/netapi/client/SaltStackClient.java [hunk 10]
```diff
@@ -290,6 +290,37 @@ public Job call() throws SaltStackException {
...
     /**
+     * Query for result of supplied job.
+     * 
+     * GET /job/<job-id>
+     * 
+     * @param job {@link Job} object representing scheduled job
+     * @return Map key: minion id, value: command result from that minion
+     * @throws SaltStackException 
+     */
+    public Map<String,Object> getJobResult( final Job job ) throws SaltStackException{
```
Comment (moio) [inline src/main/java/com/suse/saltstack/netapi/client/SaltStackClient.java]: Nitpick: don't use spaces inside parenthesis, but do use one before `{`.

File: src/main/java/com/suse/saltstack/netapi/client/SaltStackClient.java [hunk 11]
```diff
@@ -290,6 +290,37 @@ public Job call() throws SaltStackException {
...
+    	return getJobResult(job.getJid());
```
Comment (moio) [inline src/main/java/com/suse/saltstack/netapi/client/SaltStackClient.java]: Nitpick: this line is indented with a tab, please use four spaces instead.

File: src/main/java/com/suse/saltstack/netapi/client/SaltStackClient.java [hunk 12]
```diff
@@ -290,6 +290,37 @@ public Job call() throws SaltStackException {
     }
 
     /**
+     * Query for result of supplied job.
+     *
```
Comment (moio) [inline src/main/java/com/suse/saltstack/netapi/client/SaltStackClient.java]: Nitpick: there's a trailing space on this line (and others below). Please get rid of all trailing spaces.

File: src/main/java/com/suse/saltstack/netapi/client/SaltStackClient.java (hunk 1 above)
Comment (moio) [inline src/main/java/com/suse/saltstack/netapi/client/SaltStackClient.java]: I agree, if possible please correct this and other methods in a different PR.

File: src/main/java/com/suse/saltstack/netapi/client/impl/HttpClientConnection.java [hunk 13]
```diff
@@ -1,25 +1,34 @@
 package com.suse.saltstack.netapi.client.impl;
 
-import com.suse.saltstack.netapi.client.Connection;
-import com.suse.saltstack.netapi.config.ClientConfig;
-import static com.suse.saltstack.netapi.config.ClientConfig.*;
-import com.suse.saltstack.netapi.exception.SaltStackException;
-import com.suse.saltstack.netapi.parser.JsonParser;
+import static com.suse.saltstack.netapi.config.ClientConfig.PROXY_HOSTNAME;
```
Comment (moio) [inline src/main/java/com/suse/saltstack/netapi/client/impl/HttpClientConnection.java]: Nitpick: looks like you re-organized imports automatically here, with a different style than the one that was used before. Please make sure that your patches only change lines that are supposed to be changed - in general you should not reorganize imports for code that's not being changed by your contribution.

Exception of course is if you are submitting a restyling PR :-)

File: src/main/java/com/suse/saltstack/netapi/client/impl/HttpClientConnection.java [hunk 14]
```diff
@@ -77,25 +105,31 @@ public T getResult(String data) throws SaltStackException {
...
 
         try (CloseableHttpClient httpClient = httpClientBuilder.build()) {
-            // Prepare POST request
+            // Prepare request
             URI uri = config.get(URL).resolve(endpoint);
-            HttpPost httpPost = new HttpPost(uri);
-            httpPost.addHeader("Accept", "application/json");
 
-            // POST data
+            HttpUriRequest req = null;
```
Comment (moio) [inline src/main/java/com/suse/saltstack/netapi/client/impl/HttpClientConnection.java]: Nitpick: please use a more descriptive variable name here.

File: src/test/java/com/suse/saltstack/netapi/client/SaltStackClientTest.java [hunk 15]
```diff
@@ -125,9 +125,9 @@ public void testLoginAsyncFailure() throws SaltStackException {
     public void testRunRequest() throws SaltStackException {
         stubFor(post(urlEqualTo("/run"))
                 .willReturn(aResponse()
-                    .withStatus(HttpURLConnection.HTTP_OK)
-                    .withHeader("Content-Type", "application/json")
-                    .withBody(JSON_RUN_RESPONSE)));
+                        .withStatus(HttpURLConnection.HTTP_OK)
```
Comment (moio) [inline src/test/java/com/suse/saltstack/netapi/client/SaltStackClientTest.java]: Nitpick: it seems like you changed spacing here. Again please make sure that your patches only change lines that are supposed to be changed.

**Comment by sschober** [comment]
@moio 
I've tried to address all your concerns, rebased and pushed a new version.

File: src/main/java/com/suse/saltstack/netapi/client/SaltStackClient.java [hunk 16]
```diff
@@ -290,6 +290,37 @@ public Job call() throws SaltStackException {
...
+     * GET /job/<job-id>
+     *
+     * @param job String representing scheduled job
+     * @return Map key: minion id, value: command result from that minion
+     * @throws SaltStackException if anything goes wrong
+     */
+    public Map<String,Object> getJobResult(final String job) throws SaltStackException {
+        Result<List<Map<String,Object>>> result = connectionFactory
+                .create("/jobs/"+job, JsonParser.RETVALS, config)
```
Comment (renner) [inline src/main/java/com/suse/saltstack/netapi/client/SaltStackClient.java]: One more nitpick: Please put a blank before and after the `+` when concatenating strings.

**Comment by moio** [comment]
@sschober: thanks for this PR! Merged manually.

Please note I added two further nitpicks with b34c9d8 :-)

**Pull request merged**